### PR TITLE
[ MIP 1 ] Support publish acknowledgement

### PR DESCRIPTION
**Pull request by mattisonchao**
## Motivation

See #367 

## Modification

- Remove ``MqttPubExceptionHandler.java`` .
- Use builder modal refactor helper.
- Use ``AckHandler`` to process MQTT protocol PublishAck method.
- Refactor some inappropriate code.

### CODE REVIEW

**Comment by mattisonchao** [comment]
I need to fix some problem.

**Comment by mattisonchao** [comment]
@Technoboy- 
PTAL :)

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/AbstractQosPublishHandler.java [hunk 1]
```diff
@@ -51,25 +51,26 @@ protected AbstractQosPublishHandler(PulsarService pulsarService, MQTTServerConfi
...
+    /**
+     * Convert mqtt protocol message to pulsar message and send it.
+     * @param msg                    MQTT protocol message
+     * @param checkSubscriptionExist Check if the subscription exists, throw #{MQTTNoMatchingSubscriberException}
+     *                              if the subscription does not exist;
+     */
+    protected CompletableFuture<PositionImpl> writeToPulsarTopic(MqttPublishMessage msg,
+                                                                 boolean checkSubscriptionExist) {
         return getTopicReference(msg).thenCompose(topicOp -> topicOp.map(topic -> {
+            if (checkSubscriptionExist && topic.getSubscriptions().isEmpty()) {
```
Comment (Technoboy-) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/AbstractQosPublishHandler.java:66]: checkSubscriptionExist -> checkSubscription

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/DefaultProtocolMethodProcessorImpl.java [hunk 2]
```diff
@@ -162,9 +162,14 @@ public void processPublish(MqttPublishMessage msg) {
...
         result.thenAccept(__ -> msg.release())
-              .exceptionally(ex -> {
-                    log.error("[{}] Write {} to Pulsar topic failed.", msg.variableHeader().topicName(), msg, ex);
-                    msg.release();
+                .exceptionally(ex -> {
+                    Throwable cause = ex.getCause();
+                    log.error("[Publish] [{}] Write {} to Pulsar topic failed.",
+                            msg.variableHeader().topicName(), msg, cause);
+                    // Prevent after thenAccept(__ -> msg.release()) get exceptions.
+                    if (msg.refCnt() != 0) {
```
Comment (Technoboy-) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/DefaultProtocolMethodProcessorImpl.java:170]: msg.release could not throw exception, if it throws exception, we'd better use safeRelease instead. So , here. we should not add `msg.refCnt() != 0`

**Review by Technoboy- — APPROVED** [review]
LGTM, please help to check the comment.

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/DefaultProtocolMethodProcessorImpl.java (hunk 2 above)
Comment (mattisonchao) [reply mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/DefaultProtocolMethodProcessorImpl.java]: fixed

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/AbstractQosPublishHandler.java (hunk 1 above)
Comment (mattisonchao) [reply mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/AbstractQosPublishHandler.java]: fixed

**Comment by mattisonchao** [comment]
> LGTM, please help to check the comment.

Fixed, PTAL :)

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/messages/factory/MqttPubAckMessageHelper.java [hunk 3]
```diff
@@ -13,84 +13,65 @@
...
+        }
+
+        public MqttPubAckMessageHelper.MqttPubErrorAckBuilder packetId(int packetId) {
+            this.packetId = packetId;
+            return this;
+        }
+
+        public MqttPubAckMessageHelper.MqttPubErrorAckBuilder errorReason(
+                Mqtt5PubReasonCode errorReason) {
+            this.errorReason = errorReason;
```
Comment (Technoboy-) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/messages/factory/MqttPubAckMessageHelper.java:55]: errorReason -> reasonCode

File: mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/messages/factory/MqttPubAckMessageHelper.java [hunk 4]
```diff
@@ -13,84 +13,65 @@
...
-     */
-    public static MqttMessage createMqtt(int packetId) {
-        return new MqttPubAckMessage(new MqttFixedHeader(MqttMessageType.PUBACK, false, MqttQoS.AT_MOST_ONCE, false, 0),
-                MqttMessageIdVariableHeader.from(packetId));
-    }
+        public MqttPubErrorAckBuilder(int protocolVersion) {
+            this.protocolVersion = protocolVersion;
+        }
 
+        public MqttPubAckMessageHelper.MqttPubErrorAckBuilder reasonString(String reasonStr) {
```
Comment (Technoboy-) [inline mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/messages/factory/MqttPubAckMessageHelper.java:43]: reasonString -> reason

**Pull request merged**